data_safe_haven/config/config.py

```diff
@@ -1,199 +1,416 @@
+    def azure(self) -> ConfigSectionAzure:
+        if not self.azure_:
+            self.azure_ = ConfigSectionAzure()
+        return self.azure_
+
+    @property
+    def backend(self) -> ConfigSectionBackend:
+        if not self.backend_:
+            self.backend_ = ConfigSectionBackend(
+                key_vault_name=f"shm-{self.shm_name_[:9]}-kv-backend",
+                managed_identity_name=f"shm-{self.shm_name_}-identity-reader-backend",
+                resource_group_name=self.backend_resource_group_name,
+                storage_account_name=self.backend_storage_account_name,
+                storage_container_name=self.backend_storage_container_name,
             )
-            blob_client.upload_blob(str(self), overwrite=True)
-            self.logger.info(
-                f"Uploaded config [green]{self.name}[/] to blob storage.",
+        return self.backend_
+
+    @property
+    def pulumi(self) -> ConfigSectionPulumi:
+        if not self.pulumi_:
+            self.pulumi_ = ConfigSectionPulumi()
+        return self.pulumi_
+
+    @property
+    def shm(self) -> ConfigSectionSHM:
+        if not self.shm_:
+            self.shm_ = ConfigSectionSHM(name=self.shm_name_)
+        return self.shm_
+
+    @property
+    def tags(self) -> ConfigSectionTags:
+        if not self.tags_:
+            self.tags_ = ConfigSectionTags(deployment=self.name)
+        return self.tags_
+
+    def __str__(self) -> str:
+        """String representation of the Config object"""
+        contents = {}
+        if self.azure_:
+            contents["azure"] = self.azure.to_dict()
+        if self.backend_:
+            contents["backend"] = self.backend.to_dict()
+        if self.pulumi_:
+            contents["pulumi"] = self.pulumi.to_dict()
+        if self.shm_:
+            contents["shm"] = self.shm.to_dict()
+        if self.sres:
+            contents["sre"] = {k: v.to_dict() for k, v in self.sres.items()}
+        if self.tags:
+            contents["tags"] = self.tags.to_dict()
+        return str(yaml.dump(contents, indent=2))
+
+    def read_stack(self, name: str, path: pathlib.Path):
+        """Add a Pulumi stack file to config"""
+        with open(path, "r", encoding="utf-8") as f_stack:
+            b64string = f_stack.read()
+        self.pulumi.stacks[name] = b64encode(b64string)
+
+    def remove_sre(self, name: str) -> None:
+        """Remove SRE config section by name"""
+        if name in self.sres.keys():
+            del self.sres[name]
+
+    def remove_stack(self, name: str) -> None:
+        """Remove Pulumi stack section by name"""
+        if name in self.pulumi.stacks.keys():
+            del self.pulumi.stacks[name]
+
+    def write_stack(self, name: str, path: pathlib.Path):
+        """Write a Pulumi stack file from config"""
+        b64string = b64decode(self.pulumi.stacks[name])
+        with open(path, "w", encoding="utf-8") as f_stack:
+            f_stack.write(b64string)
+
+    def download(self):
```

Comment:
Would it be better to have this as a classmethod?

Reply:
Maybe. We'd have to move `_azure_api` to a class-level property too (is that a thing?). What would we gain?

Reply:
Refactored